### PR TITLE
chore: Refactor apply resources functions

### PR DESCRIPTION
**Pull request by laurentluce**
Divide by three the number of lines required to support reconciling a new kardinal-managed resource type.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -14,7 +14,6 @@ issues:
         - lll
     - path: "kardinal/*"
       linters:
-        - dupl
```
Comment (laurentluce) [inline .golangci.yml:17]: Re-enabled this linter now that those apply resources functions are not duplicated.

File: internal/controller/core/flow_controller.go [hunk 2]
```diff
@@ -55,10 +54,7 @@ type FlowReconciler struct {
 func (r *FlowReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	_ = log.FromContext(ctx)
 
-	err := reconciler.Reconcile(ctx, r.Client)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
+	_ = reconciler.Reconcile(ctx, r.Client)
```
Comment (laurentluce) [inline internal/controller/core/flow_controller.go:57]: Run the reconciler at most once for each event to help with development.  We can change the logic when needed.

**Comment by laurentluce** [comment]
The linter is crashing in CI, will debug more tomorrow morning.

File: kardinal/resources/resources.go [hunk 3]
```diff
@@ -210,179 +221,6 @@ func ApplyServiceResources(ctx context.Context, clusterResources *Resources, clu
...
-			if found && isManaged == trueStr {
-				if clusterTopologyNamespace == nil || clusterTopologyNamespace.GetDeployment(deployment.Name) == nil {
-					logrus.Infof("Deleting deployment %s", deployment.Name)
-					err := cl.Delete(ctx, deployment)
-					if err != nil {
-						return stacktrace.Propagate(err, "An error occurred deleting deployment %s", deployment.Name)
-					}
-				}
-			}
-			/* else {
```
Comment (leoporoli) [inline kardinal/resources/resources.go:253]: Is it worth moving this commented code block somewhere?

File: kardinal/resources/resources.go [hunk 4]
```diff
@@ -161,46 +162,56 @@ func AddAnnotations(obj *metav1.ObjectMeta, annotations map[string]string) {
...
-
-func ApplyServiceResources(ctx context.Context, clusterResources *Resources, clusterTopologyResources *Resources, cl client.Client) error {
+func ApplyResources(
+	ctx context.Context,
+	// Current cluster resources
+	clusterResources *Resources,
+	// Base + flows resources to reconcile
+	clusterTopologyResources *Resources,
+	cl client.Client,
+	// Function to retrieve the namespace resources
```
Comment (leoporoli) [inline kardinal/resources/resources.go]: What do you think of moving these comments to a docs comment format over the function header?
https://tip.golang.org/doc/comment#func

File: kardinal/resources/resources.go [hunk 5]
```diff
@@ -161,46 +162,56 @@ func AddAnnotations(obj *metav1.ObjectMeta, annotations map[string]string) {
...
-			for _, service := range clusterTopologyNamespace.Services {
-				namespaceService := namespace.GetService(service.Name)
-				if namespaceService == nil {
-					logrus.Infof("Creating service %s", service.Name)
-					err := cl.Create(ctx, service)
+			for _, clusterTopologyNamespaceObject := range getObjectsFunc(clusterTopologyNamespace) {
+				namespaceObject := getObjectFunc(namespace, clusterTopologyNamespaceObject.GetName())
+				if namespaceObject == nil {
+					logrus.Infof("Creating %s %s", clusterTopologyNamespaceObject.GetObjectKind().GroupVersionKind().String(), clusterTopologyNamespaceObject.GetName())
+					err := cl.Create(ctx, clusterTopologyNamespaceObject, client.FieldOwner(fieldOwner))
```
Comment (leoporoli) [inline kardinal/resources/resources.go:183]: Nice adding fieldOwner

File: kardinal/resources/resources.go [hunk 6]
```diff
@@ -161,46 +162,56 @@ func AddAnnotations(obj *metav1.ObjectMeta, annotations map[string]string) {
...
 					if err != nil {
-						return stacktrace.Propagate(err, "An error occurred creating service %s", service.Name)
+						return stacktrace.Propagate(err, "An error occurred creating resource %s", clusterTopologyNamespaceObject.GetName())
```
Comment (leoporoli) [inline kardinal/resources/resources.go]: could we include kind here too?

File: kardinal/resources/resources.go [hunk 7]
```diff
@@ -161,46 +162,56 @@ func AddAnnotations(obj *metav1.ObjectMeta, annotations map[string]string) {
...
-						if !reflect.DeepEqual(namespaceService.Spec, service.Spec) {
-							service.ResourceVersion = namespaceService.ResourceVersion
-							err := cl.Update(ctx, service)
+						if !compareObjectsFunc(clusterTopologyNamespaceObject, namespaceObject) {
+							logrus.Infof("Updating %s %s", clusterTopologyNamespaceObject.GetObjectKind().GroupVersionKind().String(), clusterTopologyNamespaceObject.GetName())
+							clusterTopologyNamespaceObject.SetResourceVersion(namespaceObject.GetResourceVersion())
+							err := cl.Update(ctx, clusterTopologyNamespaceObject)
 							if err != nil {
-								return stacktrace.Propagate(err, "An error occurred updating service %s", service.Name)
+								return stacktrace.Propagate(err, "An error occurred updating resource %s", clusterTopologyNamespaceObject.GetName())
```
Comment (leoporoli) [inline kardinal/resources/resources.go]: could we include kind here too?

File: kardinal/resources/resources.go [hunk 8]
```diff
@@ -161,46 +162,56 @@ func AddAnnotations(obj *metav1.ObjectMeta, annotations map[string]string) {
...
 			if found && isManaged == trueStr {
-				if clusterTopologyNamespace == nil || clusterTopologyNamespace.GetService(service.Name) == nil {
-					logrus.Infof("Deleting service %s", service.Name)
-					err := cl.Delete(ctx, service)
+				if clusterTopologyNamespace == nil || getObjectFunc(clusterTopologyNamespace, namespaceObject.GetName()) == nil {
+					logrus.Infof("Deleting %s %s", namespaceObject.GetObjectKind().GroupVersionKind().String(), namespaceObject.GetName())
+					err := cl.Delete(ctx, namespaceObject, client.PropagationPolicy(metav1.DeletePropagationForeground))
 					if err != nil {
-						return stacktrace.Propagate(err, "An error occurred deleting service %s", service.Name)
+						return stacktrace.Propagate(err, "An error occurred deleting resource %s", namespaceObject.GetName())
```
Comment (leoporoli) [inline kardinal/resources/resources.go]: could we include kind here too?

**Review by leoporoli — APPROVED** [review]

File: kardinal/resources/resources.go (hunk 3 above)
Comment (laurentluce) [reply kardinal/resources/resources.go:253]: I am wondering if we should set the istio inject attrbiute at the namespace level instead.  I added a `NOTES` file at the root to keep track of the component log level annotation when we need it again.

**Pull request merged**